!!! FEATURE: Separate properties from references in NodeType declaration

nezaniel opened the pull request:
Related #4549

Separates property and reference declaration by introducing a dedicated `references` section:

```yaml
'My.Package:NodeWithReferences':
   references:
     singleReference:
       constraints:
         maxItems: 1
     multiReferences: {}
```

The above is a one to one replacement for

```yaml
'My.Package:NodeWithReferences':
   properties:
     singleReference:
       type: reference
     multiReferences:
       type: references
```

**Upgrade instructions**

This pr is made backwards compatible in that the declaration of `reference` properties are still allowed and will be handled at runtime.
The nodetype contains a breaking change, as the `references` are not any longer available via `getProperties` and `getPropertyType`.
For theses usecases `getReferences` and `hasReference` were introduced.

Accessing `getPropertyType` is unsafe for references and will throw:

```php
$propertyType = $nodeType->getPropertyType($propertyName);
if ($propertyType === 'reference' || $propertyType === 'references') {
    // special stuff
}
// ...
```

The code must be changed to:

```php
if ($nodeType->hasReference($propertyName)) {
    // special stuff
    break;
}
if ($nodeType->hasProperty($propertyName) {
    $propertyType = $nodeType->getPropertyType($propertyName);
    // ...
}
```

Existing Neos 9 projects using the legacy property-like declaration syntax AND the new features like reference `constraints` or `properties` must be migrated to the new syntax with this change.

**Checklist**

Ui part: https://github.com/neos/neos-ui/pull/3668


- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the 9.0 branch
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

Commit:
4549 - separate properties from references in nodetype declaration

Commit:
Merge branch '9.0' into 4549-separateReferencesAndProperties

Commit:
4549 - WIP: separate properties and references on setFullConfiguration

Commit:
Merge branch '9.0' into 4549-separateReferencesAndProperties

Commit:
4549 - Merge references to properties for now

mhsdesign commented:
How do we want to continue here? I think i remember that this change caused some trouble in the neos ui, as we need to handle somehow the reference editors.

You proposed this syntax, which i like as it resembles this approach: https://github.com/neos/neos-development-collection/issues/4631#issuecomment-1772649612 And we already the node type constraint: `constraints.nodeTypes`

```yaml
My.NodeType:
  references:
    regularRef: {}
    singleRef:
      constraints:
        valueReference:
          maxValue: 1
```

Maybe `constraints.maxItems` or `constraints.maxReferences` would suit better but thats a detail for now.

About the validation in the commands: We can easily enforce this invariant as the command will not merge the references but replace them. Otherwise we would get into trouble when merging two workspaces. But as said we can simply assert that the to be set references are allowed by the max value.
 
----

Regarding the neos ui: We must decide whether it will be easy to implement real references and distinct between properties, or if we should still communicate towards the ui in the legacy format.

----

Something else i remember from our discussions, which i wanted to document: We want to limit in the nodeType to have either a property or a reference per name. Even-though the escr could technically handle a property and reference with the same name, it makes things in the Ui and other places more complex without real benefit. We discussed to validate this in the construction of the NodeType (`initializeNodeType`)

bwaidelich commented:
> Maybe constraints.maxItems or constraints.maxReferences

I'd prefer `maxItems`,  that's what I suggested with https://github.com/neos/neos-development-collection/issues/4631#issuecomment-1772649612 (and that was taken from JSON schema and the like)

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 4549-separateReferencesAndProperties

mhsdesign committed:
TASK: Make use of `requireNodeTypeToDeclareProperty`

mhsdesign committed:
FEATURE: Add `constraints.maxItems` for references

mhsdesign committed:
BUGFIX: Prevent to set twice the same target in a reference

mhsdesign committed:
WIP: Preparations for the Neos Ui to work again

- added `hasReference`
- __legacyPropertyType used to ensure that the FlowQuery property operation will return the node directly but not an array of nodes
- remove type "reference" or "references" in php doc documentation
- use `hasReference` instead of `getPropertyType === reference`

mhsdesign committed:
TASK: Add `NodeReferencesToWrite::createEmpty` to demonstrate use-case

mhsdesign committed:
TASK: Migrate type:references earlier before Postprocessors

That way the post processors are forced to be adjusted to work with the new format and will stop working for reference like property types

The inspector.dataTypes.references? configuration was removed as references are special types.
The `NodeReferenceConverter` was removed as it is out of use.

mhsdesign commented:
I got the ui to work again after i continued to split the properties and references. The adjustments for the ui can be found here: https://github.com/neos/neos-ui/pull/3668

mhsdesign committed:
TASK: Adjust further code to not check `getPropertyType === reference`

mhsdesign committed:
TASK: Switch back to legacy reference notation to avoid conflicts while merging

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 4549-separateReferencesAndProperties

mhsdesign commented:
While it was noble to use the new `reference` syntax in all the test files, i fear this is not the right time. With the workspace pr ahead im trying to have as little interference as possible, so i reverted that change via https://github.com/neos/neos-development-collection/pull/4677/commits/c5b96191ba052eceede4d0d3c6c08b031aea6cee. In the end as a followup, when everything is merged we can still cleanup our tests.

mhsdesign commented:
This is ready to be reviewed :D (in combination with the ui https://github.com/neos/neos-ui/pull/3668 change)
~I will update the pr description soon and also check up on the last (optional) todo.~

moved to https://github.com/neos/neos-development-collection/issues/4981

mhsdesign commented on Neos.ContentRepository.Core/Classes/NodeType/NodeType.php:
```diff
@@ -155,6 +155,29 @@ protected function buildFullConfiguration(): array
+                case 'reference':
+                    unset($propertyConfiguration['type']);
+                    $propertyConfiguration['constraints']['maxItems'] = 1;
+                    // @deprecated remove with 10
+                    // used to ensure that the FlowQuery property operation will return the node directly but not an array of nodes
+                    $propertyConfiguration['__legacyPropertyType'] = 'reference';
```
that might need to be discussed?

Reasoning is that flowquery handled multiple references and single references different in the past.
With the new syntax there is always multiple ones and maybe a limit to 1 or any other value.

To make `q(node).property('myLegacyReferenceProperty')` return the node directly but `q(node).property('referenceWithLimit1')`and array of nodes, we need this distinction.

mhsdesign replied:
For legacy we can also just look into `constraints.maxItems: 1`

mhsdesign committed:
TASK: A NodeType must have distinct names for properties and references

mhsdesign committed:
TASK: Add NodeType unit tests for new functionality

mhsdesign committed:
TASK: Limit new feature `constraints.maxItems` to new syntax

mhsdesign committed:
TASK: Limit new feature `properties` to new `references` syntax

mhsdesign committed:
TASK: Remove unnecessary check in NodeType that is done by phpstan

The constructor is @internal and thus we dont have to validate it too strongly.

mhsdesign committed:
TASK: Make NodeType `final`

mhsdesign committed:
TASK: PropertyOperation treat all 'constraints.maxItems' === 1 equally as single reference

and return the first (only) node instead of an array

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 4549-separateReferencesAndProperties

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 4549-separateReferencesAndProperties

mhsdesign committed:
TASK: Adjust to `requireNodeType`

nezaniel commented on Neos.Neos/Configuration/Settings.yaml:
```diff
@@ -230,12 +230,6 @@ Neos:
-            editor: Neos.Neos/Inspector/Editors/ReferenceEditor
-          references:
-            typeConverter: Neos\Neos\Service\Mapping\NodeReferenceConverter
-            editor: Neos.Neos/Inspector/Editors/ReferencesEditor
```
How are references converted nowadays?

mhsdesign replied:
I currently have this logic in place

```
$editor = $maxAllowedItems === 1 ? 'Neos.Neos/Inspector/Editors/ReferenceEditor' : 'Neos.Neos/Inspector/Editors/ReferencesEditor';
```

but i sure could use here the configured `dataTypes.reference` or (s) instead.
About the `typeConverter`, the ui handles references totally different in the node property conversion service see: `\Neos\Neos\Ui\Domain\Service\NodePropertyConverterService::getPropertiesArray` where `\Neos\Neos\Ui\Domain\Service\NodePropertyConverterService::convertValue` is only used for properties in Neos 9 and thus the `typeConverter` has no effect for the references.

nezaniel replied:
Ah, nice. I was just curious

nezaniel commented on Neos.ContentRepository.Core/Tests/Unit/NodeType/NodeTypeTest.php:
```diff
@@ -365,6 +391,212 @@ public function superTypesRemovedByInheritanceCanBeAddedAgain()
+        $nodeType->getReferences();
+    }
+
     /**
```
woohoo more tests <3

nezaniel commented on Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:
```diff
@@ -218,22 +221,35 @@ protected function requireNodeTypeToAllowNodesOfTypeInReference(
         }
     }
 
+    protected function requireNodeTypeToAllowCountOfReferencesInReference(SerializedNodeReferences $nodeReferences, ReferenceName $referenceName, NodeTypeName $nodeTypeName): void
```
I'd suggest NumberOf instead of CountOf, since we handle references, not votes in an election

nezaniel replied:
if only this was my own PR :D

Commit:
Rename CountOf to NumberOf

mhsdesign committed:
TASK: Make default `reference(s)` editor configurable again

mhsdesign committed:
TASK: Add references test for DefaultPropertyEditor

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into 4549-separateReferencesAndProperties

mhsdesign committed:
TASK: Simplify expression

mhsdesign reviewed (approved):
Neos ui e2e pass locally and together with the ui adjustments the node creation handle also works

Will merge this now as agreed in the weekly and with bernhard.
